### ISSUE TITLE
Exception when changing the value of a struct variable

### ISSUE THREAD

**Issue by AttilaSzobonya** [codingseb/ExpressionEvaluator#32]
Hi! I didn't do an excessive research about that bug, but I received an exception when I tried to change the value of a field of a struct.

Message:
`Operator '==' cannot be applied to operands of type '<my_struct_type>' and 'NULL'`

StackTrace:
`at CallSite.Target(Closure , CallSite , Object , Object )
at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
at CallSite.Target(Closure , CallSite , Object , Object )
at CodingSeb.ExpressionEvaluator.ExpressionEvaluator.EvaluateVarOrFunc(String expr, String restOfExpression, Stack'1 stack, Int32& i) in [...]\ExpressionEvaluator.cs:line 1833
at CodingSeb.ExpressionEvaluator.ExpressionEvaluator.Evaluate(String expression) in [...]\ExpressionEvaluator.cs:line 1385`

The expression was simply:
`struct.field = anotherVariable`

It was somehow related to the dictionary read in line 1833, where you tried to read the variable value into a dynamic variable. I was able to fix the issue by changing the type of 'cusVarValueToPush' to 'object' and only casting it to 'dynamic' in line 1872 and 1876.

I created a pull request with the fix. #31

**Comment by codingseb** [comment]
Hello and thanks for the bug report and pull request.
Until now I never used struct with ExpressionEvaluator but essencialy classes.
So there are probably some others bugs linked to that.
I will integrate your correction to next release and start some tests about the use of structs

**Comment by codingseb** [comment]
I did some tests with structs;
For now I struggling with this. Your pull request correct the specific bug you reported thanks for that.
But as struct are ValueType and are copied when given to methods and as ExpressionEvaluator use reflection (methods) to assign fields and properties, some weird things happend when you imbricate structs for example.
When you assign a value in a sub struct, the struct is copied and the assignation is never reported in the struct hierarchy.

```c#
// Set 3 on a copy of the struct2
struct1.struct2.x = 3
// This will not return 3
return struct1.struct2.x;
```

I will make more investigations. But fro now be careful when using struct with ExpressionEvaluator.

**Comment by AttilaSzobonya** [comment]
I see. : \ I spent some time digging through the code, to find a solution. I think the only way around it is that in case of a value assing we have to trace back and set every field value to the new copies.

So in your example first struct2.x = 3 (which creates a 'new struct2') then struct1.struct2 = 'new struct2'.

This would solve the problem with value types and wouldn't cause problems with reference types as you would just re-set the references.

Unfortunatelly, I don't trivially see how to achieve this in your code.

**Comment by codingseb** [comment]
Yes you are right I came to the same kind of solution trace the assignation path.
I just need now to find the way to implement it.

**Comment by codingseb** [comment]
I will try it these next couple of days.

**Comment by codingseb** [comment]
OK I found a way to manage this.

All is done this tiny class 
```c#
private class ValueTypeNestingTrace
{
	public object Container { get; set; }

	public MemberInfo Member { get; set; }

	public object Value { get; set; }

	public void AssignValue()
	{
		if (Container is ValueTypeNestingTrace valueTypeNestingTrace)
		{
			((dynamic)Member).SetValue(valueTypeNestingTrace.Value, Value);
			valueTypeNestingTrace.AssignValue();
		}
		else
		{
			((dynamic)Member).SetValue(Container, Value);
		}
	}
}
```
And some small modifications in the method `EvaluateVarOrFunc`

I did some tests and actually it works better than C# when doing something like :
```c#
classContainer = new ClassContainer()
{
    NestedStructProperty = new StructForTest1()
    {
        myIntvalue = 8,
        myStringValue = "Test"
    }
};

// Do not compile In C#
classContainer.NestedStructProperty.myIntvalue = 9;

// Return : Test 9
return $"Result {classContainer.NestedStructProperty.myStringValue} {classContainer.NestedStructProperty.myIntvalue}";
```

I don't know if it should be more like C# or if it is better like this
Maybe I will  add an option later to choose if this specific case must throw an exception.

For now I will integrate it as this in the next version

**Comment by codingseb** [comment]
Just published v : 1.3.7.0
